### PR TITLE
 Replace Pony ORM with Sqlalchemy

### LINKED ISSUE #93: Replace the usage of Pony ORM with SQLAlchemy
https://www.sqlalchemy.org/

### PR DESCRIPTION
**Pull request by calebevans**
Closes #93

### CODE REVIEW

File: qe_metrics/libs/database.py [hunk 1]
```diff
@@ -40,29 +30,30 @@ def __exit__(
...
-            password=self.db_config["password"],
-            database=self.db_config["database"],
-            port=self.db_config.get("port", 5432),
-            provider=self.db_config.get("provider", "postgres"),
-        )
+        if db_config["local"]:
+            return f"sqlite:///{db_config.get('local_filepath', '/tmp/qe_metrics.sqlite')}"
+        else:
+            verify_config(config=db_config, required_keys=["host", "user", "password", "database"])
+            drivername = db_config.get("provider", "postgresql")
```
Comment (rnetser) [inline qe_metrics/libs/database.py:53]:
```suggestion
            driver_name = db_config.get("provider", "postgresql")
```

File: tests/conftest.py [hunk 2]
```diff
@@ -45,19 +38,20 @@ def tmp_products_file(tmp_path, request):
 
 
 @pytest.fixture
-def product(db_session, tmp_sqlite_db, request):
-    product_name = request.param
-    return ProductsEntity(name=product_name)
+def product(tmp_sqlite_db, request):
```
Comment (rnetser) [inline tests/conftest.py:41]: please re-order fixture calls; first call pytest-native ones and then the ones we define

File: pyproject.toml [hunk 3]
```diff
@@ -42,11 +42,11 @@ jira = "^3.5.2"
 jinja2 = "^3.1.2"
 python-simple-logger = "^1.0.8"
 psycopg2-binary = "^2.9.9"
-pony = "^0.7.17"
 pyaml-env = "^1.2.1"
 ipdb = "^0.13.13"
 pyhelper-utils = "^0.0.9"
 flask = "^3.0.3"
+sqlalchemy = "^2.0.29"
```
Comment (rnetser) [inline pyproject.toml:49]: we can now enable python 3.12
also, please remove the comment in line 38

File: qe_metrics/utils/issue_utils.py [hunk 4]
```diff
@@ -45,17 +47,20 @@ def mark_obsolete_issues(
...
             LOGGER.info(f'Marking issue "{db_issue.issue_key}" for product {product.name} as {OBSOLETE_STR}')
             db_issue.status = OBSOLETE_STR
-    orm.commit()
+    db_session.commit()
 
 
-def update_existing_issue(existing_issue: JiraIssuesEntity, new_issue_data: Issue, severity: str) -> None:
+def update_existing_issue(
+    existing_issue: JiraIssuesEntity, new_issue_data: Issue, severity: str, db_session: Session
+) -> None:
```
Comment (rnetser) [inline qe_metrics/utils/issue_utils.py:55]: `new_issue_data` is missing from docstring
`issue` should be removed from docstring

File: qe_metrics/utils/issue_utils.py [hunk 5]
```diff
@@ -71,63 +76,73 @@ def update_existing_issue(existing_issue: JiraIssuesEntity, new_issue_data: Issu
...
         jira_server (str): Jira server URL
+        db_session (Session): SQLAlchemy Session instance.
 
     Returns:
         List["JiraIssuesEntity"]: A list of JiraIssuesEntity objects
     """
     for issue in issues:
-        if existing_issue := JiraIssuesEntity.get(issue_key=issue.key, product=product):
-            update_existing_issue(existing_issue=existing_issue, new_issue_data=issue, severity=severity)
+        if existing_issue := db_session.query(JiraIssuesEntity).filter_by(issue_key=issue.key, product=product).first():
```
Comment (rnetser) [inline qe_metrics/utils/issue_utils.py:99]: `db_session.query` returns `Query`; please fix typing

please see the ide warnings about typing (Inspection Lens is a good plugin to use)

**Review by rnetser — CHANGES_REQUESTED** [review]

File: qe_metrics/libs/database.py [hunk 6]
```diff
@@ -1,36 +1,26 @@
...
-        self.logger = get_logger(name=self.__class__.__module__)
-        orm.set_sql_debug(debug=verbose)
-        self.config_file = config_file
-        self.db_config = parse_config(path=self.config_file)["database"]
+        self.logger = get_logger(name=__name__)
+        self.connection_string = self.connection_string_builder(db_config=parse_config(path=config_file)["database"])
+        self.verbose = verbose
 
     def __enter__(self) -> "Database":
-        self.bind_local_db_connection() if self.db_config.get("local") else self.bind_remote_db_connection()
```
Comment (myakove) [inline qe_metrics/libs/database.py:32]: `__enter__` and `__exit__` is now redundant, we should use `with session` from `sqlalchemy`.

```
def __init__(self, config_file: str, verbose: bool) -> None:
        self.logger = get_logger(name=__name__)
        self.connection_string = self.connection_string_builder(db_config=parse_config(path=config_file)["database"])
        self.verbose = verbose
        self.engine = create_engine(url=self.connection_string, echo=self.verbose)
        Base.metadata.create_all(bind=self.engine)

    def session(self):
        return Session(bind=self.engine)

```

and usage:

```
db = Database(config_file=config_file, verbose=verbose_db)
with db.session() as db_session, Jira(config_file=config_file) as jira:

```

**Review by myakove — CHANGES_REQUESTED** [review]

File: qe_metrics/utils/issue_utils.py (hunk 5 above)
Comment (calebevans) [reply qe_metrics/utils/issue_utils.py:99]: @rnetser I'm a bit confused about what needs to be fixed in the typing? `db_session.query` does return `Query`, but `db_session.query.first()` returns `(JiraIssuesEntity | None)`.

Thanks for the plugin suggestion.

Comment (myakove) [reply qe_metrics/utils/issue_utils.py:99]: `first() ` return `Any`
`_T = TypeVar("_T", bound=Any)`

`    def first(self) -> Optional[_T]:`

Comment (myakove) [reply qe_metrics/utils/issue_utils.py:99]: resolving for now, if needed we will change it later.

**Pull request merged**